src/playground_app.cpp

```diff
@@ -0,0 +1,195 @@
+      .Title = spec.Title, .Width = spec.Width, .Height = spec.Height};
+  auto window = Window::create(window_spec);
+  if (!window.has_value()) {
+    std::cerr << "Window creation error" << std::endl;
+    return std::unexpected(window.error());
+  }
+  app->m_Window = std::move(*window);
+  // Set curr gl context
+  // TODO: Move this somewhere else.
+  glfwMakeContextCurrent((GLFWwindow*)app->m_Window->getNativeWindow());
+
+  int version = gladLoadGL(glfwGetProcAddress);
+  if (version == 0) {
+    std::cerr << "Failed to initialize GLAD";
+    glfwTerminate();
+    return std::unexpected(EngineError::GLAD_ERROR);
+  }
+
+  // TODO: maybe move somewhere?
+  // imgui
+  // State
+  app->m_State = {};
+  // Setup Dear ImGui context
+  IMGUI_CHECKVERSION();
+  ImGui::CreateContext();
+  ImGuiIO& io = ImGui::GetIO();
+  io.ConfigFlags |=
+      ImGuiConfigFlags_NavEnableKeyboard;  // Enable Keyboard Controls
+  io.ConfigFlags |=
+      ImGuiConfigFlags_NavEnableGamepad;  // Enable Gamepad Controls
+  io.ConfigFlags |= ImGuiConfigFlags_DockingEnable;
+
+  // Setup Platform/Renderer backends
+  ImGui_ImplGlfw_InitForOpenGL(
+      (GLFWwindow*)app->m_Window->getNativeWindow(),
+      true);  // Second param install_callback=true will install
+              // GLFW callbacks and chain to existing ones.
+  ImGui_ImplOpenGL3_Init();
+
+  return std::move(app);
+}
+
+PlaygroundApp::PlaygroundApp() : Application() {}
+
+PlaygroundApp::~PlaygroundApp() {
+  ImGui_ImplOpenGL3_Shutdown();
+  ImGui_ImplGlfw_Shutdown();
+  ImGui::DestroyContext();
+  glfwTerminate();
+}
+
+Window& PlaygroundApp::getWindow() { return *m_Window; }
+
+void PlaygroundApp::run() {
+  auto window = (GLFWwindow*)m_Window->getNativeWindow();
+
+  if (m_ActiveProgram == nullptr) {
+    std::cerr << "No active program configured!" << std::endl;
+    return;
+  }
+
+  m_ActiveProgram->setup();
+
+  // Loop
+  while (!glfwWindowShouldClose(window)) {
+    glfwPollEvents();
+    glClearColor(0.0f, 0.0f, 0.0f, 1.0f);
+    glClear(GL_COLOR_BUFFER_BIT);
+
+    ImGui_ImplOpenGL3_NewFrame();
+    ImGui_ImplGlfw_NewFrame();
+    ImGui::NewFrame();
+    showProgramSelector();
+    // Handle some state after imgui
+    // TODO: Make it event based
+    if (m_State.Run && m_State.SelectedProgram != nullptr) {
+      // Yes this is kinda dumb
+      runProgram(m_State.SelectedProgram->getName());
+      m_State.Run = false;
+    }
+
+    m_ActiveProgram->loop();
+
+    ImGui::Render();
+    ImGui_ImplOpenGL3_RenderDrawData(ImGui::GetDrawData());
+    m_Window->update();
+  }
+}
+
+void PlaygroundApp::runProgram(const std::string& name) {
+  // Why do we need 2 lookups.
+  if (m_Programs.find(name) == m_Programs.end()) {
+    std::cerr << std::format("No program {} found.", name);
+    return;
+  }
+
+  // Cleanup old program
+  if (m_ActiveProgram != nullptr) {
+    m_ActiveProgram->cleanup();
+  }
+
+  m_ActiveProgram = m_Programs.at(name).get();
+  m_ActiveProgram->setup();
```

Comment:
You don't need 2 lookups. You can just

```suggestion
  // Why do we need 2 lookups.
  if (auto program = m_Programs.find(name); 
          program == m_Programs.end()) {
    std::cerr << std::format("No program {} found.", name);
    return;
  }

  // Cleanup old program
  if (m_ActiveProgram != nullptr) {
    m_ActiveProgram->cleanup();
  }

  m_ActiveProgram = program;
  m_ActiveProgram->setup();
```

Actually program will go out of scope, you can fix that manually I guess.